samples/nrf9160/modem_shell/src/sock/sock.c

```diff
@@ -0,0 +1,1320 @@
+	 */
+	double throughput = 8 * 1000 * ((double)data_len / time_ms);
+
+	return throughput;
+}
+
+static void print_throughput_summary(uint32_t data_len, int64_t time_ms)
+{
+	double throughput = calculate_throughput(data_len, time_ms);
+
+	shell_print(
+		shell_global,
+		"Summary:\n"
+		"Data length: %7u bytes\n"
+		"Time:        %7.2f s\n"
+		"Throughput:  %7.0f bit/s",
+		data_len,
+		(float)time_ms / 1000,
+		throughput);
+}
+
+static void sock_print_data_hex(uint8_t *buffer, uint32_t buffer_size)
+{
+	/* Print received data in hexadecimal format having 8 bytes per line.
+	 * This is not made with single shell_print because we would need to
+	 * reserve a lot bigger buffer for converting all data into hexadecimal string.
+	 */
+	char hex_data[81];
+	int data_printed = 0;
+	int data_left;
+	int print_chars;
+
+	while (data_printed < buffer_size) {
+		data_left = buffer_size - data_printed;
+		print_chars = data_left <= 8 ? data_left : 8;
+
+		for (int i = 0; i < print_chars; i++) {
+			sprintf(hex_data + i * 5, "0x%02X ", buffer[data_printed + i]);
+		}
+		shell_print(shell_global, "\t%s", hex_data);
+		data_printed += print_chars;
+	}
+}
+
+static int sock_send(
+	struct sock_info_t *socket_info,
+	char *data,
+	int length,
+	bool log_data,
+	bool data_hex_format)
+{
+	int bytes;
+	int dest_addr_len = 0;
+
+	if (log_data) {
+		if (data_hex_format) {
+			shell_print(shell_global, "Socket data send:");
+			sock_print_data_hex(data, length);
+		} else {
+			shell_print(shell_global, "Socket data send:\n\t%s", data);
+		}
+	}
+
+	if (socket_info->type == SOCK_DGRAM) {
+		/* UDP */
+		if (socket_info->family == AF_INET) {
+			dest_addr_len = sizeof(struct sockaddr_in);
+		} else if (socket_info->family == AF_INET6) {
+			dest_addr_len = sizeof(struct sockaddr_in6);
+		}
+		bytes = sendto(socket_info->fd, data, length, 0,
+			       socket_info->addrinfo->ai_addr, dest_addr_len);
+	} else {
+		/* TCP and raw socket */
+		bytes = send(socket_info->fd, data, length, 0);
+	}
+	if (bytes < 0) {
+		/* Ideally we'd like to log the failure here but non-blocking
+		 * socket causes huge number of failures due to incorrectly
+		 * set POLLOUT flag:
+		 * https://devzone.nordicsemi.com/f/nordic-q-a/65392/bug-nrf9160-tcp-send-flow-control-seems-entirely-broken
+		 * Hence, we'll log only if we have blocking socket
+		 */
+		if (sock_get_blocking_mode(socket_info->fd)) {
+			shell_print(
+				shell_global,
+				"socket send failed, err %d",
+				errno);
+		}
+		return -1;
+	}
+	return bytes;
+}
+
+static void data_send_work_handler(struct k_work *item)
+{
+	struct data_transfer_info *data_send_info_ptr =
+		CONTAINER_OF(item, struct data_transfer_info, work);
+	struct sock_info_t *socket_info = data_send_info_ptr->parent;
+
+	if (!socket_info->in_use) {
+		shell_print(
+			shell_global,
+			"Socket id=%d not in use. Fatal error and sending won't work.",
+			socket_info->id);
+		/* TODO: stop timer */
```

Comment:
It should be a single instruction call (`k_timer_stop`), why is it on TODO?

Reply:
Added k_timer_stop call. I believe this branch is very hard, if not impossible, to reach but maybe there is a timing when it can occur.